### PR TITLE
[compiler-rt][ubsan][nfc-ish] Fix a type conversion bug

### PR DESCRIPTION
**Pull request by alanzhao1**
If the inline asm version of `ptrauth_strip` is used instead of the builtin, the inline asm implementation will return an unsigned long, causing an incompatible pointer conversion issue.

### CODE REVIEW

**Comment by asl** [comment]
> With #100483, if the inline asm version of `ptrauth_strip` is used instead of the builtin, the inline asm implementation will return an unsigned long, causing an incompatible pointer conversion issue.

To be more precise, it was inline asm version before that PR. It was just an implicit cast due to function call result.

**Review by asl — APPROVED** [review]
Provided that downstream builds pass. AFAIK there are no public builtbots that checks pac-ret configuration.

Thanks!

**Comment by alanzhao1** [comment]
> > With #100483, if the inline asm version of `ptrauth_strip` is used instead of the builtin, the inline asm implementation will return an unsigned long, causing an incompatible pointer conversion issue.
> 
> To be more precise, it was inline asm version before that PR. It was just an implicit cast due to function call result.

Updated the PR description.

**Comment by asl** [comment]
@alanzhao1 Actually, the spec for `ptrauth_sign` is:
```
   The result will have the same type as the original value.
```

So, I'd think we also fix the inline asm version while here. How about the following:

```
      __typeof(__value) ret;                \
```
?

**Comment by alanzhao1** [comment]
> @alanzhao1 Actually, the spec for `ptrauth_sign` is:
> 
> ```
>    The result will have the same type as the original value.
> ```
> 
> So, I think we'd also fix the inline asm version while here. How about the following:
> 
> ```
>       __typeof(__value) ret;                \
> ```
> 
> ?

Done

File: compiler-rt/lib/sanitizer_common/sanitizer_ptrauth.h [hunk 1]
```diff
@@ -27,7 +27,7 @@
           : "=r"(ret)                   \
           : "r"(__value)                \
           : "x30");                     \
-      ret;                              \
+      __typeof(__value) ret;            \
```
Comment (asl) [inline compiler-rt/lib/sanitizer_common/sanitizer_ptrauth.h:30]: Oh, no, not here :)

File: compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp [hunk 2]
```diff
@@ -207,7 +207,8 @@ struct VtablePrefix {
   std::type_info *TypeInfo;
 };
 VtablePrefix *getVtablePrefix(void *Vtable) {
-  Vtable = ptrauth_strip(Vtable, ptrauth_key_cxx_vtable_pointer);
+  Vtable = reinterpret_cast<void *>(
```
Comment (vitalybuka) [inline compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:210]: what is the purpose of this change?

File: compiler-rt/lib/sanitizer_common/sanitizer_ptrauth.h (hunk 1 above)
Comment (alanzhao1) [reply compiler-rt/lib/sanitizer_common/sanitizer_ptrauth.h]: Oops. Fixed.

File: compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp (hunk 2 above)
Comment (asl) [reply compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:210]: @alanzhao1 I think with typeof it should be redundant as return value would have `void*` type. But please double check, I do not have pac-ret enabled compiler-rt right now in buildable state...

Comment (alanzhao1) [reply compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:210]: Chrome was seeing build failures when building compiler-rt for android - see https://crbug.com/355271781#comment3.

The problem was that if we used the inline asm version of `ptrauth_strip(...)`, it woud return a `unsigned long` that would be assigned to `Vtable`, a variable of type `void *`, causing an incompatible type conversion compile failure.

This PR has been updated to fix the return value of `ptrauth_strip(...)`, but LLVM's style seem to be to explicitly cast the value to the desired type anyways.

Comment (alanzhao1) [reply compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp]: Ended up removing the `reintepret_cast` here.

**Review by vitalybuka — APPROVED** [review]

**Comment by asl** [comment]
@tru This will apply cleanly after https://github.com/llvm/llvm-project/pull/100634

**Comment by asl** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/25f9415713f9f57760a5322876906dc11385ef8e
